Allow upgrade requests in mixed content if the scheme is handled by scheme handler.

basuke opened the pull request:
#### 2408b78fe65465e82a7b3b00a80f2327938fb3f1
<pre>
Allow upgrade requests in mixed content if the scheme is handled by scheme handler.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281883">https://bugs.webkit.org/show_bug.cgi?id=281883</a>
<a href="https://rdar.apple.com/131077815">rdar://131077815</a>

Reviewed by Matthew Finkel.

In 274409@main, request upgrade in mixed content was implemented strictly following the spec.
<a href="https://www.w3.org/TR/mixed-content/#upgrade-algorithm">https://www.w3.org/TR/mixed-content/#upgrade-algorithm</a>

But for the compatibility of WKWebView, request upgrade should be allowed when the request is handled by special scheme handler.

This patch adds exception for the upgradable check in MixedContentChecker::shouldUpgradeInsecureContent().

Also adding first API test for MixedContentChecker by extracting the check for easier unit testing.

* Source/WebCore/loader/MixedContentChecker.cpp:
(WebCore::destinationIsImageAndInitiatorIsImageset):
(WebCore::MixedContentChecker::shouldUpgradeInsecureContent):
(WebCore::MixedContentChecker::canModifyRequest):
* Source/WebCore/loader/MixedContentChecker.h:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/MixedContentChecker.cpp: Added.
(TestWebKitAPI::TEST(MixedContentChecker, CanModifyRequest)):

Canonical link: <a href="https://commits.webkit.org/285709@main">https://commits.webkit.org/285709@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e7f46fb7b8f55c928124bc6c77dd333e7a5ee47

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73570 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52999 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26380 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77859 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24808 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75685 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62132 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/784 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/57838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/16245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76637 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/47973 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63302 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/38244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/44568 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/20787 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23141 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66318 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21136 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79457 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/887 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/364 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66213 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1029 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63311 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65493 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16185 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9342 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7523 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/851 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/3598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/880 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/867 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/886 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Tools/TestWebKitAPI/Tests/WebCore/MixedContentChecker.cpp:
```diff
@@ -0,0 +1,104 @@
+        destination,
+        Initiator::Imageset,
+        shouldUpgradeIPAddressAndLocalhostForTesting
+    ));
```
Can you two more tests that use a custom scheme? One test where the scheme is not registered, and a second test where the scheme is registered?

basuke replied:
Yeah! Sorry, that's the important test for the change I've made.

basuke committed:
Allow upgrade requests in mixed content if the scheme is handled by scheme handler.
https://bugs.webkit.org/show_bug.cgi?id=281883
rdar://131077815

Reviewed by Matthew Finkel.

In 274409@main, request upgrade in mixed content was implemented strictly following the spec.
https://www.w3.org/TR/mixed-content/#upgrade-algorithm

But for the compatibility of WKWebView, request upgrade should be allowed when the request is handled by special scheme handler.

This patch adds exception for the upgradable check in MixedContentChecker::shouldUpgradeInsecureContent().

Also adding first API test for MixedContentChecker by extracting the check for easier unit testing.

* Source/WebCore/loader/MixedContentChecker.cpp:
(WebCore::destinationIsImageAndInitiatorIsImageset):
(WebCore::MixedContentChecker::shouldUpgradeInsecureContent):
(WebCore::MixedContentChecker::canModifyRequest):
* Source/WebCore/loader/MixedContentChecker.h:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/MixedContentChecker.cpp: Added.
(TestWebKitAPI::TEST(MixedContentChecker, CanModifyRequest)):

Canonical link: https://commits.webkit.org/285709@main
